code/classes/DMSDocumentCart.php

```diff
@@ -0,0 +1,215 @@
+
+    /**
+     * A handle to the classes' {@link DMSCartBackendInterface}
+     *
+     * @var DMSCartBackendInterface
+     */
+    protected $backend;
+
+    public function __construct($backend = null)
+    {
+        parent::__construct();
+        $this->backend = ($backend) ?: DMSSessionBackend::create(); // Default to DMSSessionBackend if not provided
+    }
+
+    /**
+     * Returns all the cart items as an array
+     *
+     * @return array
+     */
+    public function getItems()
+    {
+        return $this->backend->getItems();
+    }
+
+    /**
+     * Add an {@link DMSRequestItem} object into the cart.
+     *
+     * @param DMSRequestItem $item
+     *
+     * @return DMSCartBackendInterface
+     */
+    public function addItem(DMSRequestItem $item)
+    {
+        return $this->backend->addItem($item);
+    }
+
+    /**
+     * Get a {@link DMSRequestItem} object from the cart.
+     *
+     * @param int $itemID The ID of the item
+     *
+     * @return DMSRequestItem|boolean
+     */
+    public function getItem($itemID)
+    {
+        return $this->backend->getItem($itemID);
+    }
+
+    /**
+     * Removes a {@link DMSRequestItem} from the cart by it's id
+     *
+     * @param DMSRequestItem $item
+     */
+    public function removeItem(DMSRequestItem $item)
+    {
+        $this->backend->removeItem($item);
+    }
+
+    /**
+     * Removes a {@link DMSRequestItem} from the cart by it's id
+     *
+     * @param int $itemID
+     */
+    public function removeItemByID($itemID)
+    {
+        $this->backend->removeItemByID($itemID);
+    }
+
+    /**
+     * Adjusts (increments, decrements or amends) the quantity of an {@link DMSRequestItem}
+     *
+     * @param int $itemID
+     * @param int $quantity
+     * @param string $updateType
+     *
+     * @return bool
+     * @throws InvalidArgumentException
+     */
+    public function updateItemQuantity($itemID, $quantity, $updateType = self::ITEM_QUANTITY_UPDATE)
+    {
+        $item = $this->getItem($itemID);
+        if (!$item) {
+            return false;
+        }
+        $currentQuantity = $item->getField('Quantity');
+        switch ($updateType) {
+            case self::ITEM_QUANTITY_UPDATE:
+                $item->setField('Quantity', $quantity);
+                $this->addItem($item);
+                break;
+            case self::ITEM_QUANTITY_ADD:
+                $item->setField('Quantity', $currentQuantity + $quantity);
+                $this->addItem($item);
+                break;
+            case self::ITEM_QUANTITY_DEDUCT:
+                $newQuantity = $currentQuantity - $quantity;
+                if ($newQuantity > 0) {
+                    $item->setField('Quantity', $newQuantity);
+                    $this->addItem($item);
+                } else {
+                    // If the quantity falls below 1, the item is removed completely.
+                    $this->removeItemByID($itemID);
+                }
+                break;
```

Comment:
What do you think about this? I think it's overly complicated for what it needs to be.

I was imagining something like this:

```php
$cart->updateItemQuantity($itemId, 2); // add two
$cart->updateItemQuantity($itemId, -1); // remove one
$cart->updateItemQuantity($itemId, 1); // add one
```

then you could just have something like this:

```php
$currentQuantity = $item->getQuantity(); // note: dog fooding the public API
$newQuantity = $currentQuantity + $increment;
if ($newQuantity < 0) {
    $newQuantity = 0;
}
$item->setQuantity($newQuantity);
return $this;
```

thoughts?